src/main/java/io/gingersnap_project/cdc/consumer/BatchConsumer.java

```diff
@@ -20,22 +26,42 @@ public class BatchConsumer implements DebeziumEngine.ChangeConsumer<ChangeEvent<
    private static final Logger log = LoggerFactory.getLogger(BatchConsumer.class);
    private final EngineWrapper engine;
    private final EventProcessingChain chain;
+   private final Executor executor;
 
-   public BatchConsumer(EngineWrapper engine, EventProcessingChain chain) {
+   public BatchConsumer(EngineWrapper engine, EventProcessingChain chain, Executor executor) {
       this.chain = chain;
       this.engine = engine;
+      this.executor = executor;
    }
 
    @Override
    public void handleBatch(List<ChangeEvent<SourceRecord, SourceRecord>> records, DebeziumEngine.RecordCommitter<ChangeEvent<SourceRecord, SourceRecord>> committer) {
       log.info("Processing {} entries", records.size());
 
       try {
-         for (ChangeEvent<SourceRecord, SourceRecord> ev : records) {
-            // TODO: we may be able to do this in parallel later or asynchronously even
-            chain.process(create(ev), new EventContext());
-            committer.markProcessed(ev);
+         Map<Object, ChangeEvent<SourceRecord, SourceRecord>> acc = new HashMap<>();
+         for (ChangeEvent<SourceRecord, SourceRecord> curr : records) {
+            acc.compute(curr.value().key(), (key, prev) -> {
+               if (prev != null) {
+                  uncheckedCommit(prev, committer);
+               }
+               return curr;
+            });
          }
+
+         committer.markBatchFinished();
```

Comment:
Do we need this twice or just after the `CompletionStages.join`?

Reply:
Nope, my mistake during rebase  :sweat_smile: :sweat_smile:

Reply:
Updated, thanks